### PR TITLE
test: base_timer

### PR DESCRIPTION
**Pull request by hongriTianqi**

### CODE REVIEW

File: source/module_base/test/timer_test.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	ModuleBase::timer::start();
+	// start() called tick() once
+	EXPECT_FALSE(ModuleBase::timer::timer_pool[""]["total"].start_flag);
+}
+
+TEST_F(TimerTest, PrintAll)
+{
+	ModuleBase::timer::tick("wavefunc","evc");
+	EXPECT_FALSE(ModuleBase::timer::timer_pool["wavefunc"]["evc"].start_flag);
+	for (int i=0;i<100;i++)	EXPECT_EQ(i,i);
```
Comment (dyzheng) [inline source/module_base/test/timer_test.cpp]: Maybe using "Sleep()" is a simple method for testing timer.

File: source/module_base/test/timer_test.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
+#include "../timer.h"
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+#include "mpi.h"
```
Comment (dyzheng) [inline source/module_base/test/timer_test.cpp]: Please add ifdef __MPI for mpi.h header

File: source/module_base/test/timer_test.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	}
+};
+
+
+TEST_F(TimerTest, Tick)
+{
+	ModuleBase::timer::tick("wavefunc","evc");
+	// after 1st call of tick, start_flag becomes false
+	EXPECT_FALSE(ModuleBase::timer::timer_pool["wavefunc"]["evc"].start_flag);
+	sleep(0.1); // ms
```
Comment (caic99) [inline source/module_base/test/timer_test.cpp]: Where is the definition of `sleep`? 
I suggest using [`sleep_for`](https://www.cplusplus.com/reference/thread/this_thread/sleep_for/).

File: source/module_base/test/timer_test.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	// then we can have time elapsed in cpu_second
+	ModuleBase::timer::tick("wavefunc","evc");
+	EXPECT_GT(ModuleBase::timer::timer_pool["wavefunc"]["evc"].cpu_second,0.0);
```
Comment (caic99) [inline source/module_base/test/timer_test.cpp]: Is `0.0` OK? Maybe we should check elapsed time being greater than the time we slept, like `0.1` here?

File: source/module_base/test/timer_test.cpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+{
+	ModuleBase::timer::start();
+	// start() called tick() once
+	EXPECT_FALSE(ModuleBase::timer::timer_pool[""]["total"].start_flag);
+}
+
+TEST_F(TimerTest, PrintAll)
+{
+	ModuleBase::timer::tick("wavefunc","evc");
+	EXPECT_FALSE(ModuleBase::timer::timer_pool["wavefunc"]["evc"].start_flag);
```
Comment (caic99) [inline source/module_base/test/timer_test.cpp]: The test of `start_flag` has already been performed; to make sure each test case only test on related part, we may not check it here. What do you think?

**Review by caic99 — CHANGES_REQUESTED** [review]

File: source/module_base/test/timer_test.cpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	getline(ifs,output);
+	getline(ifs,output);
+	getline(ifs,output);
+	getline(ifs,output);
+	EXPECT_THAT(output,testing::HasSubstr("TIME(Sec)"));
+	ifs.close();
+}
+
+// use __MPI to activate parallel environment
+int main(int argc, char **argv)
```
Comment (caic99) [inline source/module_base/test/timer_test.cpp:125]: Maybe we can wrap the whole `main()` under `#ifdef __MPI`.
Do we need to test both MPI and sequence timing? @dyzheng

Comment (dyzheng) [reply source/module_base/test/timer_test.cpp:125]: Testing for both MPI and sequence are needed.

**Review by dyzheng — APPROVED** [review]

**Comment by dyzheng** [comment]
Please add annotations for source file "timer.h" either.

**Review by caic99 — APPROVED** [review]

**Pull request merged**